lib/cookie/secureContext.ts

```diff
@@ -0,0 +1,140 @@
+  // 127.0.0.0/8: first octet = 127
+  const octets = address.split('.');
+  return (
+    octets.length === 4 &&
+    octets[0] !== undefined &&
+    parseInt(octets[0], 10) === 127
+  );
+}
+
+/**
+ * Checks if the given IPv6 address is the loopback address (`::1`).
+ *
+ * @remarks
+ * @param address - The IPv6 address to check.
+ * @returns `true` if the address is loopback, otherwise `false`.
+ */
+function isLoopbackV6(address: string): boolean {
+  // new URL(...) follows the WHATWG URL Standard
+  // which compresses IPv6 addresses, therefore the IPv6
+  // loopback address will always be compressed to '[::1]':
+  // https://url.spec.whatwg.org/#concept-ipv6-serializer
+  return (address === '::1');
+}
+
+/**
+ * Determines if the given address (IPv4 or IPv6) is a loopback address.
+ *
+ * @remarks
+ * @param address - The IP address to check.
+ * @returns `true` if the address is loopback, otherwise `false`.
+ */
+function isIpLoopback(address: string): boolean {
+  if (isIPv4(address)) {
+    return isLoopbackV4(address);
+  }
+
+  if (isIPv6(address)) {
+    return isLoopbackV6(address);
+  }
+
+  return false;
+}
+
+/**
+ * Checks if the given lowercase host ends with the `.localhost` TLD.
+ *
+ * @remarks
+ * @param lowerHost - The lowercase host string to check.
+ * @returns `true` if the host ends with `.localhost`, otherwise `false`.
+ */
+function isNormalizedLocalhostTLD(lowerHost: string): boolean {
+  return lowerHost.endsWith('.localhost');
+}
+
+/**
+ * Checks if the given host is `localhost` or matches `*.localhost`.
+ *
+ * @remarks
+ * @param host - The host string to check.
+ * @returns `true` if the host is considered local, otherwise `false`.
+ */
+function isLocalHostname(host: string): boolean {
+  const lowerHost = host.toLowerCase();
+  return lowerHost === 'localhost' ||
+    isNormalizedLocalhostTLD(lowerHost);
+}
+
+/**
+ * Removes leading and trailing square brackets if present.
+ * Adapted from https://github.com/chromium/chromium/blob/main/url/gurl.cc#L440-L448
+ *
+ * @param host - The host string to process.
+ * @returns The host string without leading and trailing square brackets if previously present.
+ */
+function hostNoBrackets(host: string): string {
+  if (host.length >= 2 && host.startsWith('[') && host.endsWith(']')) {
+    return host.substring(1, host.length - 1);
+  }
+  return host;
+}
+
+/**
+ * Determines if a URL string represents a potentially trustworthy origin.
+ * 
+ * A URL is considered potentially trustworthy if it:
+ * - Uses HTTPS or WSS schemes
+ * - Points to a loopback address (IPv4 127.0.0.0/8 or IPv6 ::1)
+ * - Uses localhost or *.localhost hostnames
+ * 
+ * @param inputUrl - The URL string or URL object to check.
+ * @returns `true` if the URL is potentially trustworthy, otherwise `false`.
+ * @see {@link https://w3c.github.io/webappsec-secure-contexts/#potentially-trustworthy-origin W3C Spec}
+ */
+export function isPotentiallyTrustworthy(inputUrl: string | URL): boolean {
```

Comment:
```suggestion
export function isSecureConnection(inputUrl: string | URL): boolean {
```

[RFC6265bis-19](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-19) uses the terminology `"secure" connection` when referring to the behaviour being outlined in this PR. E.g.;

> If the request-uri does not denote a "secure" connection (as defined by the user agent), and the cookie's secure-only-flag is true, then abort these steps and ignore the cookie entirely.
> - https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-19#section-5.7-3.13.1

> NOTE: The notion of a "secure" connection is not defined by this document. Typically, user agents consider a connection secure if the connection makes use of transport-layer security, such as SSL or TLS, or if the host is trusted. For example, most user agents consider "https" to be a scheme that denotes a secure protocol and "localhost" to be trusted host.
> - https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-19#section-5.8.3-2.1.2.3.2

This function should reflect that terminology and be renamed to `isSecureConnection`.

Reply:
[Secure Contexts](https://www.w3.org/TR/secure-contexts/) uses the terminology 'potentially trustworthy'. What you described is an actual secure connection, i.e. if the web server has an SSL certificate installed.

When a browser (i.e. Chromium) decides to send `Secure` cookies, it checks the scheme for `https`, `wss`, etc. If it then connects to the server and it doesn't have an SSL certificate the request will fail anyways thus the cookies won't get sent.

`localhost` is an exception to this as it is considered potentially trustworthy.

A potentially trustworthy origin indicates that the scheme is of secure nature and/or we are on localhost. An actual "secure" connection means the web server has an SSL certificate which we can't check.

Therefore I believe `isPotentiallyTrustworthy` is more suitable here as it would be misleading to label it as secure when we in fact don't know that.

Reply:
Understood @Chriss4123, though my intention here was to ground the naming in language referenced in [RFC6265bis-19](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-19) which is deliberately vague about how it's defining this concept. Again, I'll point to the following:

> NOTE: The notion of a "secure" connection is not defined by this document. Typically, user agents consider a connection secure if the connection makes use of transport-layer security, such as SSL or TLS, or if the host is trusted. For example, most user agents consider "https" to be a scheme that denotes a secure protocol and "localhost" to be trusted host.
> - https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-19#section-5.8.3-2.1.2.3.2

Based on the examples given in the above text, the concept of a "secure connection" from this RFC can cover both the scheme and host. 

You point out in your reply, that this is inaccurate and the terminology from [Secure Contexts](https://www.w3.org/TR/secure-contexts/) is more suitable. I don't disagree with you but I still prefer the terminology of "secure connection" to be used in the code because:
- In [RFC6265bis-19](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-19) you will find the phrase `"secure" connection` referenced at multiple points in the implementation details in both the Retrieval Algorithm and the Storage Model where the code from your PR is called
- The phrase `potentially trustworthy origin` can be found buried in [Appendix A of RFC6265bis-19](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-19#appendix-A-1.7.1) but there's no mention of [Secure Contexts](https://www.w3.org/TR/secure-contexts/) in the list of references. 

So I would ask you to consider this from the perspective of a maintainer. It's easier to compare the implementation to the RFC when the language used in both matches. Even if that language is vague and somewhat misleading (which the RFC explicitly calls out) 😅  

Mind you, I'm not going to die on this hill. There's enough overlap in the terminology to backup both arguments (https://github.com/httpwg/http-extensions/issues/2605, [httpwg/http-extensions#2759](http://github.com/httpwg/http-extensions/pull/2759)). If we keep the naming of `isPotentiallyTrustworthy` then some extra comments will need to be added to connect the dots between how [Secure Contexts](https://www.w3.org/TR/secure-contexts/) relates to the implementation details of [RFC6265bis-19](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-19).

Reply:
> Understood @Chriss4123, though my intention here was to ground the naming in language referenced in [RFC6265bis-19](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-19) which is deliberately vague about how it's defining this concept. Again, I'll point to the following:
> 
> > NOTE: The notion of a "secure" connection is not defined by this document. Typically, user agents consider a connection secure if the connection makes use of transport-layer security, such as SSL or TLS, or if the host is trusted. For example, most user agents consider "https" to be a scheme that denotes a secure protocol and "localhost" to be trusted host.
> > 
> > * https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-19#section-5.8.3-2.1.2.3.2
> 
> Based on the examples given in the above text, the concept of a "secure connection" from this RFC can cover both the scheme and host.
> 
> You point out in your reply, that this is inaccurate and the terminology from [Secure Contexts](https://www.w3.org/TR/secure-contexts/) is more suitable. I don't disagree with you but I still prefer the terminology of "secure connection" to be used in the code because:
> 
>     * In [RFC6265bis-19](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-19) you will find the phrase `"secure" connection` referenced at multiple points in the implementation details in both the Retrieval Algorithm and the Storage Model where the code from your PR is called
> 
>     * The phrase `potentially trustworthy origin` can be found buried in [Appendix A of RFC6265bis-19](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-19#appendix-A-1.7.1) but there's no mention of [Secure Contexts](https://www.w3.org/TR/secure-contexts/) in the list of references.
> 
> 
> So I would ask you to consider this from the perspective of a maintainer. It's easier to compare the implementation to the RFC when the language used in both matches. Even if that language is vague and somewhat misleading (which the RFC explicitly calls out) 😅
> 
> Mind you, I'm not going to die on this hill. There's enough overlap in the terminology to backup both arguments ([httpwg/http-extensions#2605](https://github.com/httpwg/http-extensions/issues/2605), [httpwg/http-extensions#2759](https://github.com/httpwg/http-extensions/pull/2759)). If we keep the naming of `isPotentiallyTrustworthy` then some extra comments will need to be added to connect the dots between how [Secure Contexts](https://www.w3.org/TR/secure-contexts/) relates to the implementation details of [RFC6265bis-19](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-19).

 RFC6265bis-19 deliberately avoids rigidly defining what constitutes a “secure” connection for cookies, instead deferring that decision to user agents. In the spec’s [definition](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis#section-4.1.2.5) of the `Secure` attribute, it explicitly notes that **“’secure’ is defined by the user agent”**. The cookie [retrieval](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis#section-5.8.3) algorithm also requires that a `Secure` cookie be sent only if the request’s URL is over a “secure” channel – again **left to the UA’s definition**. An informative note in the spec:

> *“NOTE: The notion of a ‘secure’ connection is not defined by this document. Typically, user agents consider a connection secure if the connection makes use of transport-layer security, such as SSL or TLS, or if the host is trusted. For example, most user agents consider ‘https’ to be a scheme that denotes a secure protocol and ‘localhost’ to be [a] trusted host.”* ([draft-ietf-httpbis-rfc6265bis-19](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis#name-retrieval-algorithm))

This makes it clear the IETF spec intentionally refrains from hard-coding a definition of “secure connection.” Instead, it acknowledges common UA behavior (treating TLS-protected schemes as secure, and even treating certain **trusted hosts like `localhost` as secure-equivalent**), but leaves the exact criteria up to implementations.

RFC6265bis-19 expects browsers to apply their own notion of what is secure. Modern browsers have, in practice, converged on the concept of **“potentially trustworthy origins”** to make this determination, rather than a naive check of scheme or transport alone.

## Potentially Trustworthy Origins
Broadly, the phrase *“secure connection”* is less precisely defined, whereas **“potentially trustworthy origin”** (and the related concept of **“secure contexts”**) is the well-defined, normative standard used to gate security-sensitive behaviors. Multiple authoritative specifications define and use the concept of “potentially trustworthy” in place of ad-hoc notions of security:

- **W3C Secure Contexts Specification:** This spec defines what it means for an origin to be *potentially trustworthy*. It provides an [algorithm](https://www.w3.org/TR/secure-contexts/#is-origin-trustworthy) enumerating conditions under which an origin is considered secure enough. The Secure Contexts spec’s [core definition](https://www.w3.org/TR/secure-contexts/#is-origin-trustworthy) is: *“A potentially trustworthy origin is one which a user agent can generally trust as delivering data securely.”* This formalizes the intuitive idea mentioned in RFC6265bis’s note [that `https` and `localhost` are considered secure](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis#name-retrieval-algorithm) into a normative algorithm.

- **WHATWG HTML Standard:** HTML integrates the secure contexts concept to determine when a browsing context is a “secure context.” The HTML spec states that *“an environment is a secure context if the following algorithm returns true…”*, and one of the key steps in that [algorithm](https://html.spec.whatwg.org/multipage/webappapis.html#secure-contexts) is: *“If the result of **Is URL potentially trustworthy?** given *environment’s top-level creation URL* is ‘Potentially Trustworthy’, then return true.”*. This defers to the Secure Contexts spec’s [“Is origin potentially trustworthy?”](https://w3c.github.io/webappsec-secure-contexts/#potentially-trustworthy-url) check for the top-level page’s origin to decide if a context is secure.

- **WHATWG Fetch Standard:** The Fetch spec similarly uses “potentially trustworthy” in defining network request behaviors. For example, the algorithms for handling Mixed Content use this terminology: *“Upgrade a mixed content request to a potentially trustworthy URL, if appropriate.”* ([Fetch Standard - WhatWG](https://fetch.spec.whatwg.org/#main-fetch)).

Summarized: **Secure Contexts, HTML, and Fetch (along with related specs like Mixed Content) all treat “potentially trustworthy origin” as the canonical definition of a secure environment.** This covers HTTPS and WSS, but also explicitly includes other cases like `localhost`, and excludes others (plain `http` to remote hosts) – thus providing a precise security model. The term *“secure connection”* by itself has no formal definition in these specs and could be interpreted in narrower ways (e.g. “only TLS”).

## Browser Implementations
Browser behavior has evolved to align with the **“potentially trustworthy origin”** model as it’s more accurate and future-proof than a simplistic “secure connection” check. Both Chromium and Firefox have made conscious changes to use the broader *trustworthy origin* concept for cookies marked Secure, rather than just checking for HTTPS:

- **Firefox:** Firefox historically allowed Secure cookies on `https://` but not on `http://` origins. However, as developers increasingly run local servers on `http://localhost` and expect to use Secure cookies (or the `__Host-`/`__Secure-` name prefixes which require secure contexts), Mozilla updated Firefox’s implementation to treat **`http://localhost` as secure for cookies**. The Firefox bug report states the plan: *“By RFC6265, the definition of ’secure’ [protocol] is defined by the user-agent. I propose to unify the implementations and use `nsMixedContentBlocker::IsPotentiallyTrustworthyOrigin()` instead of a simple ’https’ scheme check.”* ([1618113 - Allow ’secure’ cookies when set by localhost](https://bugzilla.mozilla.org/show_bug.cgi?id=1618113)). The code for this can be found [here](https://github.com/mozilla/gecko-dev/blob/af5960c7041d895174ea6bceb1822773e0c9b350/dom/security/nsMixedContentBlocker.cpp#L298-#L366).

- **Chromium:** Historically, Chrome **did** allow `Secure` cookies on `http://localhost` (treating localhost as secure), but it initially did **not** allow the stricter `__Host-` and `__Secure-` cookie name prefixes on localhost. The Chromium team has broadened the definition of “secure” for cookies to include *potentially trustworthy origins* as well. A Chromium [issue discussion](https://issues.chromium.org/issues/40196122) notes that they consider support for `__Host-`/`__Secure-` on localhost **“in scope”**. The cookie handling code defines an enumeration of cookie source schemes with categories for **cryptographic vs. non-cryptographic but trustworthy**. Chrome’s [`CookieAccessScheme`](https://chromium.googlesource.com/chromium/src/+/HEAD/net/cookies/cookie_constants.h#L265) includes: `kCryptographic` (for HTTPS/WSS, implicitly trustworthy) and `kTrustworthy` (for non-HTTPS schemes that are deemed potentially trustworthy). A [comment](https://chromium.googlesource.com/chromium/src/+/HEAD/net/cookies/cookie_constants.h#L264) in the source links directly to the W3C Secure Contexts definition of *potentially trustworthy origin*, showing that Chrome’s engineers explicitly built the **Secure=“potentially trustworthy”** logic into their implementation...

It’s clear that “secure connection” as a term is too simplistic to capture the real conditions under which cookies (and other features) should be considered secure. A strict reading might equate *secure connection* only with **TLS** however both Firefox and Chrome purposefully include **“localhost” (no TLS) as secure** based on the reasoning that it’s a trustworthy origin. The term *potentially trustworthy origin* **precisely describes the broader notion of security** – it includes *both* cryptographic transport *and* other trustworthy circumstances (like loopback interfaces, local files, or extension contexts).

## Terminology and Maintainability
The phrase “secure connection” is ambiguous without an explicit definition. The [RFC6265bis](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis#section-5.8.3-2.1.2.3.2) text itself had to add a clarifying note to hint at meaning. If our implementation simply says “if secure connection, then …” we’d likely need a similar comment explaining what “secure” entails. By using the term *“potentially trustworthy origin”* (or referring to a *secure context*), we immediately invoke a well-defined concept. It’s unambiguous to a web platform engineer that this means the origin satisfies the standard [Secure Contexts](https://www.w3.org/TR/secure-contexts/#is-origin-trustworthy) criteria. I believe future contributors are more likely to be familiar with the Secure Contexts definition (since it’s used across many specs and APIs) than with the cookie spec’s isolated wording.

Most modern web platform code (both in specs and in browser source) has standardized on talking about “trustworthy” or “secure contexts.” Our cookie implementation logic should speak the same language, aligning itself conceptually with web platforms, and making the code self-documenting in a clearer way.

It’s important to note that using *“potentially trustworthy origin”* in our implementation detail is **not actually contradicting RFC6265bis-19** – it’s clarifying it. The RFC’s intent was clearly that user agents decide what “secure” means. Today’s consensus is that *secure = potentially trustworthy origin*. [Issue discussions](https://github.com/httpwg/http-extensions/issues/2605) in the IETF HTTP Working Group suggest exactly this alignment: to *“consider reusing the notion of ‘Is origin potentially trustworthy?’ from the Secure Contexts spec, instead of [a] user-agent-dependent ‘secure protocol’ check”*. So while our implementation detail might use a different phrasing than the raw RFC text, it actually **reflects the RFC’s prescribed approach (deferring to UA judgment) more faithfully** by naming the precise criterion we’ve chosen to follow.

Sticking slavishly to the term “secure connection” for the sake of mirroring the RFC introduces ambiguity, require redundant explanatory comments, and risk drifting from the rest of the web platform’s security model.

## Conclusion 
**Using “potentially trustworthy” in place of “secure connection”** in our RFC6265bis-19 cookie jar implementation is a technically sound decision that is inline with the reality of web security today. The RFC deliberately empowers user agents to decide what “secure” means; adopting the well-defined *potentially trustworthy origin* standard is the best interpretation of that mandate.

I can make appropriate comments linking the two terms thus maintaining fidelity to RFC6265bis-19 and simultaneously ensure our implementation is clear, consistent, and maintainable for the future. Adopting *“potentially trustworthy origin”* terminology will strengthen both the technical correctness and the clarity of our cookie security model, which is an unequivocal win for people reading the code and implementers.

Reply:
Sorry about the essay, I honestly just looked at all my research I did before making this change and wrote about all of it and when I realized it was too long it was too late as I was going out that night so I though "oh well, it is what it is" :P